src/core/components/pin.js

```diff
@@ -0,0 +1,529 @@
+'use strict'
+
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const CID = require('cids')
+const pinSet = require('./pin-set')
+const KeySet = require('./key-set')
+const promisify = require('promisify-es6')
+const multihashes = require('multihashes')
+const Key = require('interface-datastore').Key
+const _ = require('lodash')
+const each = require('async/each')
+const waterfall = require('async/waterfall')
+const until = require('async/until')
+const once = require('once')
+
+const keyString = multihashes.toB58String
+
+module.exports = function pin (self) {
+  let directPins = new KeySet()
+  let recursivePins = new KeySet()
+  let internalPins = new KeySet()
+
+  // temp - should use /local/pins to be consistent with go repo, but
+  // the pin set in datastore in test/go-ipfs-repo doesn't deserialize
+  const pinDataStoreKey = new Key('/local/pins/js')
+
+  const repo = self._repo
+  const dag = self.dag
+
+  function normalizeHashes (hashes, callback) {
+    // try to accept a variety of hash options including
+    // multihash Buffers, base58 strings, and ipfs path
+    // strings, either individually or as an array
+    if (Buffer.isBuffer(hashes) || typeof hashes.forEach !== 'function') {
```

Comment:
maybe use `Array.isArray()` here?

Reply:
Hard to remember now, but I think I had used that originally and changed it after someone asked to avoid using it... Or maybe it had been `instanceof Array`. Another option is `hashes.constructor === Array`. I'm fine with any of these. Currently in the project I see three examples of `instanceof Array` and three of `Array.isArray()`.